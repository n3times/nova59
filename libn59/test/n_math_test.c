#include "n_test.h"

#include <stdio.h>

typedef n_t (*fun_t)(n_t, n_err_t *err);
typedef n_t (*fun2_t)(n_t);

static void p(n_t n, n_t res, n_err_t err, char *op) {
  char s[N_PRINT_MAX_SIZE];
  char s_res[N_PRINT_MAX_SIZE];
  printf("%s %s   =   %s%s\n",
         n_print(n, s),
         op,
         n_print(res, s_res),
         err ? "?" : "");
}

int main() {
  n_err_t err;
  n_t n;
  n_t ns[] = { TEST_NUMBERS };

  fun_t funs[] = { n_square, n_sqrt, n_log, n_ln, n_1_over_x, n_exp, n_pow10 };
  char *ops[] = { "square", "sqrt", "log", "ln", "1/x" , "exp", "pow10" };

  for (int k = 0; k < N_ELEMS(funs); k++) {
    for (int i = 0; i < N_ELEMS(ns); i++) {
      n = funs[k](ns[i], &err);
      p(ns[i], n, err, ops[k]);
    }
    printf("\n\n\n");
  }

  fun2_t fun2s[] = { n_int, n_frac, n_abs, n_chs };
  char *ops2[] = { "int", "frac", "abs", "chs" };

  for (int k = 0; k < N_ELEMS(fun2s); k++) {
    for (int i = 0; i < N_ELEMS(ns); i++) {
      n = fun2s[k](ns[i]);
      p(ns[i], n, false, ops2[k]);
    }
    printf("\n\n\n");
  }

}
