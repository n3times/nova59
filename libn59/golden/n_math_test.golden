 0000000000000 00 square   =    0000000000000 00
 1000000000000 00 square   =    1000000000000 00
 3141592653590 00 square   =    9869604401090 00
 1000000000000-99 square   =    1000000000000-99?
 9999999999999 99 square   =    9999999999999 99?
-1000000000000 00 square   =    1000000000000 00
-3141592653590 00 square   =    9869604401090 00
-1000000000000-99 square   =    1000000000000-99?
-9999999999999 99 square   =    9999999999999 99?



 0000000000000 00 sqrt   =    0000000000000 00
 1000000000000 00 sqrt   =    1000000000000 00
 3141592653590 00 sqrt   =    1772453850906 00
 1000000000000-99 sqrt   =    3162277660168-50
 9999999999999 99 sqrt   =    1000000000000 50
-1000000000000 00 sqrt   =    1000000000000 00?
-3141592653590 00 sqrt   =    1772453850906 00?
-1000000000000-99 sqrt   =    3162277660168-50?
-9999999999999 99 sqrt   =    1000000000000 50?



 0000000000000 00 log   =   -9999999999999 99?
 1000000000000 00 log   =    0000000000000 00
 3141592653590 00 log   =    4971498726942-01
 1000000000000-99 log   =   -9900000000000 01
 9999999999999 99 log   =    1000000000000 02
-1000000000000 00 log   =    0000000000000 00?
-3141592653590 00 log   =    4971498726942-01?
-1000000000000-99 log   =   -9900000000000 01?
-9999999999999 99 log   =    1000000000000 02?



 0000000000000 00 ln   =   -9999999999999 99?
 1000000000000 00 ln   =    0000000000000 00
 3141592653590 00 ln   =    1144729885849 00
 1000000000000-99 ln   =   -2279559242064 02
 9999999999999 99 ln   =    2302585092994 02
-1000000000000 00 ln   =    0000000000000 00?
-3141592653590 00 ln   =    1144729885849 00?
-1000000000000-99 ln   =   -2279559242064 02?
-9999999999999 99 ln   =    2302585092994 02?



 0000000000000 00 1/x   =    9999999999999 99?
 1000000000000 00 1/x   =    1000000000000 00
 3141592653590 00 1/x   =    3183098861838-01
 1000000000000-99 1/x   =    1000000000000 99
 9999999999999 99 1/x   =    1000000000000-99?
-1000000000000 00 1/x   =   -1000000000000 00
-3141592653590 00 1/x   =   -3183098861838-01
-1000000000000-99 1/x   =   -1000000000000 99
-9999999999999 99 1/x   =   -1000000000000-99?



 0000000000000 00 exp   =    1000000000000 00
 1000000000000 00 exp   =    2718281828459 00
 3141592653590 00 exp   =    2314069263278 01
 1000000000000-99 exp   =    1000000000000 00
 9999999999999 99 exp   =    9999999999999 99?
-1000000000000 00 exp   =    3678794411714-01
-3141592653590 00 exp   =    4321391826377-02
-1000000000000-99 exp   =    1000000000000 00
-9999999999999 99 exp   =    1000000000000-99?



 0000000000000 00 pow10   =    1000000000000 00
 1000000000000 00 pow10   =    1000000000000 01
 3141592653590 00 pow10   =    1385455731368 03
 1000000000000-99 pow10   =    1000000000000 00
 9999999999999 99 pow10   =    9999999999999 99?
-1000000000000 00 pow10   =    1000000000000-01
-3141592653590 00 pow10   =    7217841590744-04
-1000000000000-99 pow10   =    1000000000000 00
-9999999999999 99 pow10   =    1000000000000-99?



 0000000000000 00 int   =    0000000000000 00
 1000000000000 00 int   =    1000000000000 00
 3141592653590 00 int   =    3000000000000 00
 1000000000000-99 int   =    0000000000000 00
 9999999999999 99 int   =    9999999999999 99
-1000000000000 00 int   =   -1000000000000 00
-3141592653590 00 int   =   -3000000000000 00
-1000000000000-99 int   =    0000000000000 00
-9999999999999 99 int   =   -9999999999999 99



 0000000000000 00 frac   =    0000000000000 00
 1000000000000 00 frac   =    0000000000000 00
 3141592653590 00 frac   =    1415926535900-01
 1000000000000-99 frac   =    1000000000000-99
 9999999999999 99 frac   =    0000000000000 00
-1000000000000 00 frac   =    0000000000000 00
-3141592653590 00 frac   =   -1415926535900-01
-1000000000000-99 frac   =   -1000000000000-99
-9999999999999 99 frac   =    0000000000000 00



 0000000000000 00 abs   =    0000000000000 00
 1000000000000 00 abs   =    1000000000000 00
 3141592653590 00 abs   =    3141592653590 00
 1000000000000-99 abs   =    1000000000000-99
 9999999999999 99 abs   =    9999999999999 99
-1000000000000 00 abs   =    1000000000000 00
-3141592653590 00 abs   =    3141592653590 00
-1000000000000-99 abs   =    1000000000000-99
-9999999999999 99 abs   =    9999999999999 99



 0000000000000 00 chs   =    0000000000000 00
 1000000000000 00 chs   =   -1000000000000 00
 3141592653590 00 chs   =   -3141592653590 00
 1000000000000-99 chs   =   -1000000000000-99
 9999999999999 99 chs   =   -9999999999999 99
-1000000000000 00 chs   =    1000000000000 00
-3141592653590 00 chs   =    3141592653590 00
-1000000000000-99 chs   =    1000000000000-99
-9999999999999 99 chs   =    9999999999999 99



 1000000000000-01 int   =    0000000000000 00
 1000000000000 00 int   =    1000000000000 00
 1000000000000 01 int   =    1000000000000 01
 1000000000000 02 int   =    1000000000000 02
 1000000000000 03 int   =    1000000000000 03
 1000000000000 04 int   =    1000000000000 04
 1000000000000 05 int   =    1000000000000 05
 1000000000000 06 int   =    1000000000000 06
 1000000000000 07 int   =    1000000000000 07
 1000000000000 08 int   =    1000000000000 08
 1000000000000 09 int   =    1000000000000 09
 1000000000000 10 int   =    1000000000000 10
 1000000000000 11 int   =    1000000000000 11
 1000000000000 12 int   =    1000000000000 12
 1000000000000 13 int   =    1000000000000 13
 9999999999999-02 int   =    0000000000000 00
 9999999999999-01 int   =    0000000000000 00
 9999999999999 00 int   =    9000000000000 00
 9999999999999 01 int   =    9900000000000 01
 9999999999999 02 int   =    9990000000000 02
 9999999999999 03 int   =    9999000000000 03
 9999999999999 04 int   =    9999900000000 04
 9999999999999 05 int   =    9999990000000 05
 9999999999999 06 int   =    9999999000000 06
 9999999999999 07 int   =    9999999900000 07
 9999999999999 08 int   =    9999999990000 08
 9999999999999 09 int   =    9999999999000 09
 9999999999999 10 int   =    9999999999900 10
 9999999999999 11 int   =    9999999999990 11
 9999999999999 12 int   =    9999999999999 12
 9999999999999 13 int   =    9999999999999 13
 1234567890123-01 int   =    0000000000000 00
 1234567890123 00 int   =    1000000000000 00
 1234567890123 01 int   =    1200000000000 01
 1234567890123 02 int   =    1230000000000 02
 1234567890123 03 int   =    1234000000000 03
 1234567890123 04 int   =    1234500000000 04
 1234567890123 05 int   =    1234560000000 05
 1234567890123 06 int   =    1234567000000 06
 1234567890123 07 int   =    1234567800000 07
 1234567890123 08 int   =    1234567890000 08
 1234567890123 09 int   =    1234567890000 09
 1234567890123 10 int   =    1234567890100 10
 1234567890123 11 int   =    1234567890120 11
 1234567890123 12 int   =    1234567890123 12
 1234567890123 13 int   =    1234567890123 13
-1234567890123-01 int   =    0000000000000 00
-1234567890123 00 int   =   -1000000000000 00
-1234567890123 01 int   =   -1200000000000 01
-1234567890123 02 int   =   -1230000000000 02
-1234567890123 03 int   =   -1234000000000 03
-1234567890123 04 int   =   -1234500000000 04
-1234567890123 05 int   =   -1234560000000 05
-1234567890123 06 int   =   -1234567000000 06
-1234567890123 07 int   =   -1234567800000 07
-1234567890123 08 int   =   -1234567890000 08
-1234567890123 09 int   =   -1234567890000 09
-1234567890123 10 int   =   -1234567890100 10
-1234567890123 11 int   =   -1234567890120 11
-1234567890123 12 int   =   -1234567890123 12
-1234567890123 13 int   =   -1234567890123 13

 1000000000000-01 frac   =    1000000000000-01
 1000000000000 00 frac   =    0000000000000 00
 1000000000000 01 frac   =    0000000000000 00
 1000000000000 02 frac   =    0000000000000 00
 1000000000000 03 frac   =    0000000000000 00
 1000000000000 04 frac   =    0000000000000 00
 1000000000000 05 frac   =    0000000000000 00
 1000000000000 06 frac   =    0000000000000 00
 1000000000000 07 frac   =    0000000000000 00
 1000000000000 08 frac   =    0000000000000 00
 1000000000000 09 frac   =    0000000000000 00
 1000000000000 10 frac   =    0000000000000 00
 1000000000000 11 frac   =    0000000000000 00
 1000000000000 12 frac   =    0000000000000 00
 1000000000000 13 frac   =    0000000000000 00
 9999999999999-02 frac   =    9999999999999-02
 9999999999999-01 frac   =    9999999999999-01
 9999999999999 00 frac   =    9999999999990-01
 9999999999999 01 frac   =    9999999999900-01
 9999999999999 02 frac   =    9999999999000-01
 9999999999999 03 frac   =    9999999990000-01
 9999999999999 04 frac   =    9999999900000-01
 9999999999999 05 frac   =    9999999000000-01
 9999999999999 06 frac   =    9999990000000-01
 9999999999999 07 frac   =    9999900000000-01
 9999999999999 08 frac   =    9999000000000-01
 9999999999999 09 frac   =    9990000000000-01
 9999999999999 10 frac   =    9900000000000-01
 9999999999999 11 frac   =    9000000000000-01
 9999999999999 12 frac   =    0000000000000 00
 9999999999999 13 frac   =    0000000000000 00
 1234567890123-01 frac   =    1234567890123-01
 1234567890123 00 frac   =    2345678901230-01
 1234567890123 01 frac   =    3456789012300-01
 1234567890123 02 frac   =    4567890123000-01
 1234567890123 03 frac   =    5678901230000-01
 1234567890123 04 frac   =    6789012300000-01
 1234567890123 05 frac   =    7890123000000-01
 1234567890123 06 frac   =    8901230000000-01
 1234567890123 07 frac   =    9012300000000-01
 1234567890123 08 frac   =    1230000000000-02
 1234567890123 09 frac   =    1230000000000-01
 1234567890123 10 frac   =    2300000000000-01
 1234567890123 11 frac   =    3000000000000-01
 1234567890123 12 frac   =    0000000000000 00
 1234567890123 13 frac   =    0000000000000 00
-1234567890123-01 frac   =   -1234567890123-01
-1234567890123 00 frac   =   -2345678901230-01
-1234567890123 01 frac   =   -3456789012300-01
-1234567890123 02 frac   =   -4567890123000-01
-1234567890123 03 frac   =   -5678901230000-01
-1234567890123 04 frac   =   -6789012300000-01
-1234567890123 05 frac   =   -7890123000000-01
-1234567890123 06 frac   =   -8901230000000-01
-1234567890123 07 frac   =   -9012300000000-01
-1234567890123 08 frac   =   -1230000000000-02
-1234567890123 09 frac   =   -1230000000000-01
-1234567890123 10 frac   =   -2300000000000-01
-1234567890123 11 frac   =   -3000000000000-01
-1234567890123 12 frac   =    0000000000000 00
-1234567890123 13 frac   =    0000000000000 00
